wandb/sdk/wandb_run.py

```diff
@@ -1467,32 +1467,54 @@ def _datatypes_callback(self, fname: str) -> None:
         files: FilesDict = dict(files=[(GlobStr(glob.escape(fname)), "now")])
         self._backend.interface.publish_files(files)
 
-    def _serialize_custom_charts(self, data: dict[str, Any]) -> dict[str, Any]:
+    def _serialize_custom_charts(
+        self,
+        data: dict[str, Any],
+        key_prefix: str | None = None,
+    ) -> dict[str, Any]:
+        """Transform plot objects into tables, and adds the vega spec to the run config.
+
+        Args:
+            data: Dictionary containing data that may include plot objects
+                Plot objects can be nested in dictionaries, which will be processed recursively.
+
+        Returns:
+            The processed dictionary with custom charts transformed into tables.
+        """
         if not data:
             return data
 
-        chart_keys = set()
+        keys_to_replace = set()
         for k, v in data.items():
+            key = f"{key_prefix}.{k}" if key_prefix else k
+
             if isinstance(v, Visualize):
                 data[k] = v.table
-                v.set_key(k)
+                v.set_key(key)
                 self._config_callback(
                     val=v.spec.config_value,
                     key=v.spec.config_key,
                 )
             elif isinstance(v, CustomChart):
-                chart_keys.add(k)
-                v.set_key(k)
+                # If this is a custom chart, we will update our history with the table key.
+                # Allowing for charts to be nested in dictionaries.
+                keys_to_replace.add(k)
+                v.set_key(key)
                 self._config_callback(
-                    key=v.spec.config_key,
                     val=v.spec.config_value,
+                    key=v.spec.config_key,
                 )
+            elif isinstance(v, dict):
+                # Recursively apply the visualization hack to nested dictionaries
+                data[k] = self._serialize_custom_charts(v, key_prefix=key)
 
-        for k in chart_keys:
-            # remove the chart key from the row
+        for k in keys_to_replace:
+            # Remove the custom chart keys from the history.
             v = data.pop(k)
             if isinstance(v, CustomChart):
-                data[v.spec.table_key] = v.table
+                # Update our history with the table key.
+                # Only the last part of the key is needed when displaying the table.
+                data[v.spec.table_key.split(".")[-1]] = v.table
```

Comment:
This is actually very strange, and I dont think I fully understand the exact process now.

The way it appeared to me was that the table key in the vega spec was `level1.level2.level3_table` (from my script above)
and the table was mapped under the nested structure in the run summary (see screenshots above).

However this didnt seem to be working properly in all cases. After a lot of experimentation the only way I found to get it to work properly as you mentioned was to flatten the tables into a single key.
So instead of 
```json
{
  "level1": {
    "level2": {
      "level3": {wandb.plot...}
    }
  }
}
```

this gets transformed to 
```json
{
  "level1.level2.level3": {wandb.plot...}
}
```

Reply:
But why is the latter a problem? Aren't we the ones setting `custom_chart_object.table_key` to `outer.inner_table` via `set_key` calls? Why not set it to something else?

Consider: is `v.spec.table_key.split(".")[-1]` not equivalent to `key + "_table"`? I wouldn't feel comfortable using the latter, as it assumes the relationship between `CustomChartSpec.key` and `CustomChartSpec.table_key`. But so does the other approach!

Side note: have you looked into the `CustomChartSpec.split_table` field? It looks like it changes how `table_key` relates to `key`.

My suggestion is to avoid parsing `table_key`. You'd need two phases: first, pop out all the charts; then, insert each chart's table at its `table_key`.

Reply:
Actually, you are right after coming back to this with a better understanding of how chart displays and runs work together. We can directly just set the table key with the current key since we are replacing all table keys at the current level in the dictionary. So we do not need to keep track of the prefix itself.

Reply:
Discussed offline, the cleanest easiest way to handle this is by gathering all the charts, then using the table key as the new key in the data dictionary at the top level.